pymc/func_utils.py

```diff
@@ -0,0 +1,138 @@
+        PyMC distribution you want to set a prior on.
+        Needs to have a ``logcdf`` method implemented in PyMC.
+    lower : float
+        Lower bound to get `mass` % of probability of `pm_dist`.
+    upper : float
+        Upper bound to get `mass` % of probability of `pm_dist`.
+    init_guess: Dict[str, float]
+        Initial guess for ``scipy.optimize.least_squares`` to find the
+        optimal parameters of `pm_dist` fitting the interval constraint.
+        Must be a dictionary with the name of the PyMC distribution's
+        parameter as keys and the initial guess as values.
+    mass: float, default to 0.95
+        Share of the probability mass we want between ``lower`` and ``upper``.
+        Defaults to 95%.
+    fixed_params: Dict[str, float], Optional, default None
+        Only used when `pm_dist` has at least three parameters.
+        Dictionary of fixed parameters, so that there are only 2 to optimize.
+        For instance, for a StudenT, you fix nu to a constant and get the optimized
+        mu and sigma.
+
+    Returns
+    -------
+    The optimized distribution parameters as a dictionary with the parameters'
+    name as key and the optimized value as value.
+    """
+    assert 0.01 <= mass <= 0.99, (
+        "This function optimizes the mass of the given distribution +/- "
+        f"1%, so `mass` has to be between 0.01 and 0.99. You provided {mass}."
+    )
+
+    # exit when any parameter is not scalar:
+    if np.any(np.asarray(distribution.rv_op.ndims_params) != 0):
+        raise NotImplementedError(
+            "`pm.find_constrained_prior` does not work with non-scalar parameters yet.\n"
+            "Feel free to open a pull request on PyMC repo if you really need this feature."
+        )
+
+    dist_params = aet.vector("dist_params")
+    params_to_optim = {
+        arg_name: dist_params[i] for arg_name, i in zip(init_guess.keys(), range(len(init_guess)))
+    }
+
+    if fixed_params is not None:
+        params_to_optim.update(fixed_params)
+
+    dist = distribution.dist(**params_to_optim)
+
+    try:
+        logcdf_lower = pm.logcdf(dist, pm.floatX(lower))
+        logcdf_upper = pm.logcdf(dist, pm.floatX(upper))
+    except AttributeError:
+        raise AttributeError(
+            f"You cannot use `find_constrained_prior` with {distribution} -- it doesn't have a logcdf "
+            "method yet.\nOpen an issue or, even better, a pull request on PyMC repo if you really "
+            "need it."
+        )
+
+    cdf_error = (pm.math.exp(logcdf_upper) - pm.math.exp(logcdf_lower)) - mass
+    cdf_error_fn = pm.aesaraf.compile_pymc([dist_params], cdf_error, allow_input_downcast=True)
+
+    try:
+        aesara_jac = pm.gradient(cdf_error, [dist_params])
+        jac = pm.aesaraf.compile_pymc([dist_params], aesara_jac, allow_input_downcast=True)
+    # when PyMC cannot compute the gradient
+    # TODO: use specific gradient, not implemented exception
+    except Exception:
```

Comment:
I think the exception we want here is the one found in `aesara.gradient.NullTypeGradError`

Reply:
The exception that's thrown is a `TypeError` by the Aesara `grad` method (line 501) in `aesara/gradient.py`:
```python
if cost is not None and cost.ndim != 0:
>           raise TypeError("cost must be a scalar.")
```
I added that error in the Except clause and tests pass locally. `aesara.gradient.NullTypeGradError` and `NotImplementedError` don't seem to be raised but I kept them in case they are by other cases we may have forgotten

Reply:
We shouldn't catch that error. That means we produced a wrong input to aesara grad. 

The other two exceptions are the ones that appear when a grad is not implemented for an Op

Reply:
That means we should make these two exceptions appear then, shouldn't we? Because they are not raised right now -- only the TypeError is raised

Reply:
> That means we should make these two exceptions appear then, shouldn't we? Because they are not raised right now -- only the TypeError is raised
> (here is to my first GH comment of the year 🥂 )

Those two exceptions mean there is no grad implemented for some Op in the cdf which can very well happen and its a good reason to silently default to the scipy approximation. The TypeError, on the other hand, should not be catched, as that means we did something wrong. 

In fact there was a point during this PR when it was always silently defaulting to the scipy because we were passing two values to grad and suppressing the TypeError.

Reply:
Ooooh ok, I see, thanks. So you're saying that the problem we have with Exponential here is because our code is wrong, not only because the grad is failing in PyMC?

Reply:
I think the exponential is a different issue altogether. Or are you getting a TypeError during compilation for the exponential tests?

Reply:
Ah no it's for the Poisson that I get the TypeError. It's because we don't have a gradient for this one, right? If yes, I would expect one of the two other errors to be raised (or, here, silently default to Scipy)

Reply:
Okay, so that probably means there is some bug in an Op used by the Poisson logcdf

Reply:
Okay, so it seems the Poisson logcdf is not returning a scalar variable, and that was the error you were seeing

Reply:
```python
import pymc as pm
pm.logcdf(pm.Poisson.dist(1), 1).eval()
# array([-0.30685282])
```

Reply:
#5307 should fix it!

Reply:
I wonder if using the symbolic gradient gets rid of the numerical issues we were seeing in float32

Reply:
Lol, such a silly error. 
> I wonder if using the symbolic gradient gets rid of the numerical issues we were seeing in float32

That sounds promising! How would you do that?

Reply:
> Lol, such a silly error.
> 
> > I wonder if using the symbolic gradient gets rid of the numerical issues we were seeing in float32
> 
> That sounds promising! How would you do that?

Check if the Poisson test passes with the original precision of 1e-5 instead of the 1e-2

Reply:
Locally seems to be passing in float64/float32, so we don't need to have the separate test just for the Poisson anymore.

Reply:
Ha ha damn, I just pushed without that change. Tests are indeed passing locally. Gonna refactor the tests and push again

Reply:
Hot damn, tests are passing locally 🔥 Pushed!
Why does the symbolic gradien help so much with numerical errors?

Reply:
Because the logcdf uses gammaincc whose gradient is notoriously tricky. We somewhat recently added a numerically stable(r) implementation to Aesara

Reply:
https://github.com/aesara-devs/aesara/pull/513